app/gui2/src/components/ComponentBrowser/__tests__/placement.test.ts

```diff
@@ -0,0 +1,653 @@
+<script setup>
+const rects = [
+${rects
+  .filter((rect) => !selectedRects.includes(rect))
+  .map(
+    (rect) =>
+      `  { left: ${rect.left}, top: ${rect.top}, width: (${rect.right} - ${rect.left}), height: (${rect.bottom} - ${rect.top}) },`,
+  )
+  .join('\n')}
+]
+
+const selectedRects = [
+${selectedRects
+  .map(
+    (rect) =>
+      `  { left: ${rect.left}, top: ${rect.top}, width: (${rect.right} - ${rect.left}), height: (${rect.bottom} - ${rect.top}) },`,
+  )
+  .join('\n')}
+]
+
+const rect = { left: ${newNode.pos.x}, top: ${newNode.pos.y}, width: ${newNode.size.x}, height: ${
+      newNode.size.y
+    } }
+</script>
+
+<template>
+<div style="height: 1000px; width: 2000px;">
+  <div v-for="rect in rects" :style="{ position: 'absolute', left: rect.left + 'px', top: rect.top + 'px', width: rect.width + 'px', height: rect.height + 'px', background: '#88888822' }"></div>
+  <div v-for="rect in selectedRects" :style="{ position: 'absolute', left: rect.left + 'px', top: rect.top + 'px', width: rect.width + 'px', height: rect.height + 'px', background: '#4444aa44' }"></div>
+  <div :style="{ position: 'absolute', left: rect.left + 'px', top: rect.top + 'px', width: rect.width + 'px', height: rect.height + 'px', background: '#88222288' }"></div>
+  </div>
+</template>
+`
+  }
+
+  const getMousePosition = vi.fn(() => Vec2.Zero)
+  // Center is at (1100, 700)
+  const screenBounds = defaultScreenBounds
+  // Half of this is (50, 10)
+  const nodeSize = defaultNodeSize
+  function environment(nodeRects: Iterable<Rect>, selectedNodeRects: Iterable<Rect>): Environment {
+    return {
+      screenBounds,
+      nodeRects,
+      selectedNodeRects,
+      get mousePosition() {
+        return getMousePosition()
+      },
+    }
+  }
+  function environment2(nodeRects: Rect[], selectedNodeRectsStart: number): Environment {
+    return {
+      screenBounds,
+      nodeRects,
+      selectedNodeRects: nodeRects.slice(selectedNodeRectsStart),
+      get mousePosition() {
+        return getMousePosition()
+      },
+    }
+  }
+  // Previous node dictated placement MUST fail if there is no previous node.
+  expect(() => previousNodeDictatedPlacement(nodeSize, environment([], []))).toThrow()
+  // Single node
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2([new Rect(new Vec2(1050, 690), defaultNodeSize)], -1),
+    ),
+    // 20px existing node height + 24px gap
+  ).toEqual<Placement>({ position: new Vec2(1050, 734) })
+  // Single node (far enough up that it does not overlap)
+  expect(
+    previousNodeDictatedPlacement(
+      nodeSize,
+      environment2(
+        [
+          new Rect(new Vec2(1150, 714), defaultNodeSize),
+          new Rect(new Vec2(1050, 690), defaultNodeSize),
+        ],
+        -1,
+      ),
+    ),
+  ).toEqual<Placement>({ position: new Vec2(1050, 734) })
+  // Single node (far enough down that it does not overlap)
+  expect(
```

Comment:
Use [`test.each`](https://vitest.dev/api/#test-each) to reduce the boilerplate in this test suite, and also split it into multiple units. As it stands, the formatting of each `expect` is now taking the majority of space, making it hard to read.